File: app/lib/api/history_api.dart
import '../api/api_client.dart';
import '../models/course_offering.dart';
import '../models/instructor_summary.dart';
import '../models/student_summary.dart';

class HistoryApi {
  HistoryApi(this._client);

  final ApiClient _client;

  Future<List<CourseOffering>> fetchCourseOfferings(String courseCode) async {
    final code = courseCode.trim().toUpperCase();
    final data = await _client.requestJson('/api/courses/$code/offerings');
    final list = data['offerings'] as List<dynamic>? ?? [];
    return list.whereType<Map<String, dynamic>>().map(CourseOffering.fromJson).toList();
  }

  Future<List<InstructorSummary>> searchInstructors(String query) async {
    final q = query.trim();
    if (q.length < 2) return [];
    final data = await _client.requestJson(
      '/api/instructors/search',
      query: {'q': q},
    );
    final list = data['results'] as List<dynamic>? ?? [];
    return list.whereType<Map<String, dynamic>>().map(InstructorSummary.fromJson).toList();
  }

  Future<InstructorOfferings> fetchInstructorOfferings(String email) async {
    final encoded = Uri.encodeComponent(email.trim().toLowerCase());
    final data = await _client.requestJson('/api/instructors/$encoded/offerings');
    return InstructorOfferings.fromJson(data);
  }

  Future<List<StudentSummary>> searchStudents(String query) async {
    final q = query.trim();
    if (q.length < 2) return [];
    final data = await _client.requestJson(
      '/api/students/search',
      query: {'q': q},
    );
    final list = data['results'] as List<dynamic>? ?? [];
    return list.whereType<Map<String, dynamic>>().map(StudentSummary.fromJson).toList();
  }

  Future<StudentOfferings> fetchStudentOfferings(String kerberos) async {
    final encoded = Uri.encodeComponent(kerberos.trim().toLowerCase());
    final data = await _client.requestJson('/api/students/$encoded/offerings');
    return StudentOfferings.fromJson(data);
  }
}

File: src/utils/historyApi.js
import { apiFetch } from '../auth/AuthContext';

export function offeringToCourse(offering) {
    if (!offering) return null;
    return {
        courseCode: offering.courseCode,
        courseName: offering.courseName || offering.courseCode,
        semesterCode: offering.semesterCode,
        totalCredits: offering.credits ?? 0,
        creditStructure: offering.creditStructure || '0.0-0.0-0.0',
        instructor: offering.instructor || '',
        instructorEmail: offering.instructorEmail || null,
        currentStrength: offering.currentStrength || null,
        slot: {
            name: offering.slotName || null,
            lectureTimingStr: offering.lectureTimingStr || null,
            lectureTiming: offering.lectureTiming || null,
            tutorialTiming: offering.tutorialTiming || null,
            labTiming: offering.labTiming || null,
        },
        lectureHall: offering.lectureHall || null,
        offeredThisSemester: Boolean(offering.isActive),
    };
}

/** Latest semester first (YYTT numeric code decreases with time). */
export function compareSemesterCodeDesc(a, b) {
    const sa = ((a && typeof a === 'object' ? a.semesterCode : a) || '').trim();
    const sb = ((b && typeof b === 'object' ? b.semesterCode : b) || '').trim();
    if (/^\d{4}$/.test(sa) && /^\d{4}$/.test(sb)) {
        return Number(sb) - Number(sa);
    }
    return sb.localeCompare(sa);
}

/** Latest semester first; ties broken by course code. */
export function sortOfferingsBySemesterDesc(offerings) {
    return [...offerings].sort((a, b) => {
        const bySemester = compareSemesterCodeDesc(a, b);
        if (bySemester !== 0) return bySemester;
        return (a.courseCode || '').localeCompare(b.courseCode || '');
    });
}

export async function fetchSemesters() {
    const res = await apiFetch('/api/semesters');
    if (!res.ok) throw new Error('semesters_load_failed');
    const data = await res.json();
    return Array.isArray(data.semesters) ? data.semesters : [];
}

export async function fetchCourseOfferings(courseCode) {
    const code = encodeURIComponent((courseCode || '').trim().toUpperCase());
    const res = await apiFetch(`/api/courses/${code}/offerings`);
    if (!res.ok) throw new Error('offerings_load_failed');
    const data = await res.json();
    const offerings = Array.isArray(data.offerings) ? data.offerings : [];
    return {
        courseCode: data.courseCode || code,
        offerings: sortOfferingsBySemesterDesc(offerings),
    };
}

export async function searchInstructors(query) {
    const q = (query || '').trim();
    if (q.length < 2) return [];
    const res = await apiFetch(`/api/instructors/search?q=${encodeURIComponent(q)}`);
    if (!res.ok) throw new Error('instructor_search_failed');
    const data = await res.json();
    return Array.isArray(data.results) ? data.results : [];
}

export async function fetchInstructorOfferings(email) {
    const enc = encodeURIComponent((email || '').trim().toLowerCase());
    const res = await apiFetch(`/api/instructors/${enc}/offerings`);
    if (!res.ok) throw new Error('instructor_offerings_failed');
    const data = await res.json();
    if (Array.isArray(data.offerings)) {
        data.offerings = sortOfferingsBySemesterDesc(data.offerings);
    }
    return data;
}

export async function searchStudents(query) {
    const q = (query || '').trim();
    if (q.length < 2) return [];
    const res = await apiFetch(`/api/students/search?q=${encodeURIComponent(q)}`);
    if (!res.ok) throw new Error('student_search_failed');
    const data = await res.json();
    return Array.isArray(data.results) ? data.results : [];
}

export async function fetchStudentOfferings(kerberos) {
    const enc = encodeURIComponent((kerberos || '').trim().toLowerCase());
    const res = await apiFetch(`/api/students/${enc}/offerings`);
    if (!res.ok) throw new Error('student_offerings_failed');
    const data = await res.json();
    if (Array.isArray(data.offerings)) {
        data.offerings = sortOfferingsBySemesterDesc(data.offerings);
    }
    return data;
}

export function offeringTimingSummary(offering) {
    if (offering.lectureTimingStr && offering.lectureTimingStr.trim()) {
        return offering.lectureTimingStr.trim();
    }
    return '—';
}
